src/utils/stream.ts

```diff
@@ -0,0 +1,399 @@
+          batchGenerationPromise = batchGenerationPromise.then(async () => {
+            await generateNextBatch();
+          });
+        }, 0);
+      }
+
+      /**
+       * Send event to update UI.
+       * This should be throttled to avoid flooding the UI with events.
+       * The UI will be updated in intervals of at least `UI_UPDATE_INTERVAL`.
+       */
+      const updateUI = async () => {
+        if (abortController.signal.aborted) {
+          return;
+        }
+        if (canUpdateUI) {
+          canUpdateUI = false;
+          connection.emit('wallet-load-partial-update', {
+            addressesFound: await storage.store.addressCount(),
+            historyLength: await storage.store.historyCount(),
+          });
+          setTimeout(() => {
+            canUpdateUI = true;
+          }, UI_UPDATE_INTERVAL);
+        }
+      };
+
+      // This listener will handle all the messages coming from the fullnode.
+      const listener = (wsData: IStreamSyncHistoryData) => {
+        // Early return if the stream is aborted
+        // This will prevent any tx or address from being added when we want to stop the stream.
+        if (abortController.signal.aborted) {
+          return;
+        }
+        // Only process the message if it is from our stream, this error should not happen.
+        if (wsData.id !== streamId) {
+          // Check that the stream id is the same we sent
+          console.log(`Wrong stream ${JSON.stringify(wsData)}`);
+          return;
+        }
+        // Vertex is a transaction in the history of the last address received
+        if (isStreamSyncHistoryVertex(wsData)) {
+          foundAnyTx = true;
+          // add to history
+          executionQueue = executionQueue.then(async () => {
+            await storage.addTx(wsData.data);
+            await updateUI();
+          });
+        }
+        if (isStreamSyncHistoryAddress(wsData)) {
+          if (wsData.index > lastReceivedIndex) {
+            lastReceivedIndex = wsData.index;
+          }
+          // Register address on storage
+          // The address will be subscribed on the server side
+          executionQueue = executionQueue.then(async () => {
+            const alreadyExists = await storage.isAddressMine(wsData.address);
+            if (!alreadyExists) {
+              await storage.saveAddress({
+                base58: wsData.address,
+                bip32AddressIndex: wsData.index,
+              });
+            }
+            await updateUI();
+          });
+          // Generate next batch if needed
+          batchGenerationPromise = batchGenerationPromise.then(async () => {
+            await generateNextBatch();
+          });
+        }
+        if (isStreamSyncHistoryEnd(wsData)) {
+          // cleanup and stop the method.
+          // console.log(`Stream end at ${Date.now()}`);
+          // This will resolve the promise and clean the listener on the connection
+          abortController.abort();
```

Comment:
Why would you abort if the streamer signaled that it had been successfully completed?

Reply:
Changed to only resolve.

The idea was that the abort controller has some other responsabilities like stopping the client from generating more addresses and removing the event listener to avoid memory leaks.
We already have a call to the abort method on the finally so we can just resolve the promise, and let the cleanup be done when the function ends.

Reply:
Got it. Great! I'm wondering whether just StreamController would be a better name since it manages other stuff.

Reply:
I can rename the variable it but [AbortController](https://developer.mozilla.org/en-US/docs/Web/API/AbortController) is from javascript so we can't rename the class.

Reply:
Got it. Ok. Thanks for trying.